google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -0,0 +1,199 @@
+                continue
+            if self.mode == scanner_rules.RuleMode.WHITELIST:
+                if is_whitelist_violation(self.verify_rules, policy):
+                    yield self._create_violation(
+                        [policy], 'FIREWALL_WHITELIST_VIOLATION')
+            if self.mode == scanner_rules.RuleMode.BLACKLIST:
+                if is_blacklist_violation(self.verify_rules, policy):
+                    yield self._create_violation(
+                        [policy], 'FIREWALL_BLACKLIST_VIOLATION')
+
+    def _create_violation(self, policies, violation_type):
+        """Creates a RuleViolation.
+
+        Args:
+          policies (list): A list of FirewallRule that violate the policy.
+          violation_type (str): The type of violation.
+
+        Returns:
+          RuleViolation: A RuleViolation for the given policies.
+
+        Raises:
+          ValueError: If no policies are passed in.
+        """
+        if not policies:
+            raise ValueError('No policies in violation')
+        return self.RuleViolation(
+            resource_type='firewall_policy',
+            resource_id=policies[0].project_id,
+            rule_name=self.name,
+            violation_type=violation_type,
+            policy_names=[p.name for p in policies])
+
+
+    # Rule violation.
+    # resource_type: string
+    # resource_id: string
+    # rule_name: string
+    # rule_index: int
+    # violation_type: FIREWALL_VIOLATION
+    # policy_name: string
+    RuleViolation = namedtuple('RuleViolation',
+                               ['resource_type', 'resource_id', 'rule_name',
+                                'violation_type', 'policy_names'])
+# pylint: enable=too-many-instance-attributes
+
+
+def is_whitelist_violation(rules, policy):
+    """Checks if the policy is not a subset of those allowed by the rules.
+
+    Args:
+      rules (list): A list of FirewallRule that the policy must be a subset of.
+      policy (FirweallRule): A FirewallRule.
+
+    Returns:
+      bool: If the policy is a subset of one of the allowed rules or not.
+    """
+    return not any([policy < rule for rule in rules])
+
+def is_blacklist_violation(rules, policy):
+    """Checks if the policy is a superset of any not allowed by the rules.
+
+    Args:
+      rules (list): A list of FirewallRule that the policy must be a subset of.
+      policy (FirweallRule): A FirewallRule.
+
+    Returns:
+      bool: If the policy is a superset of one of the blacklisted rules or not.
+    """
+    return any([policy > rule for rule in rules])
+
+def is_rule_exists_violation(rule, policies):
+    """Checks if the rule is the same as one of the policies.
+
+    Args:
+      rule (FirweallRule): A FirewallRule.
+      policies (list): A list of FirewallRule that must have the rule.
+
+    Returns:
+      bool: If the required rule is in the policies.
+    """
+    return not any([policy == rule for policy in policies])
```

Comment:
Is == correct here, or would is_equivalent be better? Or maybe matches should use ==, but required should use is_equivalent?

Reply:
I'm not committed either way, but I was thinking it was similar to a "matches" policy in that we want a specific policy to be applied. I think if we wanted to audit for certain things being allowed but not a specific policy then we would want a separate violation type for that.